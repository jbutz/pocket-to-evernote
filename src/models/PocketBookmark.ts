export class PocketBookmark {
    public readonly href: string;
    public readonly title: string;
    public readonly timeAdded: Date;
    public readonly tags: string[];
    constructor(props: {
        href: string,
        title: string,
        timeAdded: string,
        tags?: string
    }) {
        this.href = props.href;
        this.title = props.title;
        this.timeAdded = new Date(Number.parseInt(props.timeAdded, 10) * 1000);
        this.tags = (props.tags || '').split(',').filter((value) => !!value);
    }
}