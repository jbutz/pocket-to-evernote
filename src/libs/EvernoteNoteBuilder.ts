import { PocketBookmark } from '../models/PocketBookmark';
import { create as createXml, fragment as createXmlFragment } from 'xmlbuilder2';
import { XMLBuilder } from 'xmlbuilder2/lib/interfaces';

export class EvernoteNoteBuilder {
    public build(pocketBookmark: PocketBookmark): string {
        const xmlObj = this.constructBookmarkNote(pocketBookmark);

        return xmlObj.end({prettyPrint: true});
    }

    private constructBookmarkNote(pocketBookmark: PocketBookmark): XMLBuilder {
        return createXml({ version: '1.0', encoding: 'UTF-8' })
            .ele('en-note')
            .dtd({ sysID: 'http://xml.evernote.com/pub/enml2.dtd' })
            .ele('div', { 'style': '--en-clipped-content:bookmark' })
            .ele('div', { style: 'font-size: 16px; display:block; min-width: 100%; ' })
            .ele('div')
            .import(this.constructNoteBody(pocketBookmark));
    }

    private constructNoteBody(pocketBookmark: PocketBookmark) {
        const bodyWrapper = createXmlFragment()
            .ele('div', { style: 'display:block;height:100%;' });
        bodyWrapper.import(this.constructBookmarkTitle(pocketBookmark));
        bodyWrapper.import(this.constructBookmarkBody(pocketBookmark));

        return bodyWrapper;
    }

    private constructBookmarkTitle(pocketBookmark: PocketBookmark) {
        return createXmlFragment()
            .ele('div', { 'style': 'font-size:14px;line-height:20px;font-weight:bold;padding-bottom:10px;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;', })
            .txt(pocketBookmark.title);
    }

    private constructBookmarkBody(pocketBookmark: PocketBookmark) {
        const bodyFlexContainer = createXmlFragment()
            .ele('div', { style: 'display:flex;flex-flow:row nowrap;place-content:stretch flex-start;align-items:flex-start;padding-top:15px;border-top:1px solid rgb(211, 214, 216);height:100%;' })
            .com('<en-media type="image/png" hash="ff26897593310722d6f07e6fcba357ef" width="146" height="125" style="max-width:180px;vertical-align:top;display:inline-block;"></en-media>');
        const flexContent = bodyFlexContainer.ele('div', { style: 'margin-left:15px;flex:1 1 auto;display:block;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;height:100%;' });
        flexContent.ele('div', { style: 'display:block;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;margin-bottom:15px;' })
            .com('<en-media type="image/png" hash="346e09471362f2907510a31812129cd2" width="16" height="16" style="vertical-align:top;display:inline-block;max-width:16px;"></en-media>')
            .ele('a', { href: pocketBookmark.href, target: '_blank', style: 'margin-left:10px;' })
            .txt(pocketBookmark.href);
        flexContent.ele('div', { style: 'white-space:normal;font-size:12px;overflow:hidden;height:200px;line-height:150%;' })
            .txt(pocketBookmark.title);
        return bodyFlexContainer;
    }


}