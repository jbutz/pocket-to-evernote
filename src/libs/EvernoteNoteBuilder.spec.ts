import { PocketBookmark } from "../models/PocketBookmark";
import { EvernoteNoteBuilder } from "./EvernoteNoteBuilder";
import {promises as fs} from 'fs';
import path from 'path';

describe('EvernoteNoteBuilder', () => {
    let builder: EvernoteNoteBuilder;
    beforeEach(() => {
        builder = new EvernoteNoteBuilder();
    });

    it('builds the xml document', async () => {
        const expectedXml = await fs.readFile(path.join(__dirname, '..','testing','EvernoteNoteBuilder.output00.xml'), {encoding: 'utf-8'});
        const exampleBookmark = new PocketBookmark({
            href: 'https://example.com',
            title: 'Example Bookmark',
            timeAdded: '0',
            tags: ''
        });

        expect(builder.build(exampleBookmark)).toEqual(expectedXml);
    })
});