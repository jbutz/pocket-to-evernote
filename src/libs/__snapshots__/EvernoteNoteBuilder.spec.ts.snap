// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvernoteNoteBuilder builds the xml document 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE en-note SYSTEM \\"http://xml.evernote.com/pub/enml2.dtd\\">
<en-note>
  <div style=\\"--en-clipped-content:bookmark\\">
    <div style=\\"font-size: 16px; display:block; min-width: 100%; \\">
      <div>
        <div style=\\"display:block;height:100%;\\">
          <div style=\\"font-size:14px;line-height:20px;font-weight:bold;padding-bottom:10px;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;\\">Example Bookmark</div>
          <div style=\\"display:flex;flex-flow:row nowrap;place-content:stretch flex-start;align-items:flex-start;padding-top:15px;border-top:1px solid rgb(211, 214, 216);height:100%;\\">
            <!--<en-media type=\\"image/png\\" hash=\\"ff26897593310722d6f07e6fcba357ef\\" width=\\"146\\" height=\\"125\\" style=\\"max-width:180px;vertical-align:top;display:inline-block;\\"></en-media>-->
            <div style=\\"margin-left:15px;flex:1 1 auto;display:block;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;height:100%;\\">
              <div style=\\"display:block;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;margin-bottom:15px;\\">
                <!--<en-media type=\\"image/png\\" hash=\\"346e09471362f2907510a31812129cd2\\" width=\\"16\\" height=\\"16\\" style=\\"vertical-align:top;display:inline-block;max-width:16px;\\"></en-media>-->
                <a href=\\"https://example.com\\" target=\\"_blank\\" style=\\"margin-left:10px;\\">https://example.com</a>
              </div>
              <div style=\\"white-space:normal;font-size:12px;overflow:hidden;height:200px;line-height:150%;\\">Example Bookmark</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</en-note>"
`;
